Support k-NN similarity functions in painless scripting

VijayanB opened the pull request:
*Issue #213

Implement IndexFieldDataBuilder to allow painless scripting to indenitfy KNNScriptDocValues
Add dependency to PainlessExtension to add custom scoring methods to be exposed for usage by users inside painless scripting.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

VijayanB committed:
Add IndexFieldDataBuilder

Painless scripting uses IndexFieldDataBuilder to load field data values. Hence, override
fielddataBuilder to return instance of KNNVectorIndexFieldData.Builder.
Subsequently, this builder is responsible for returning ScriptDocValues which
will be used as argument in our Whitlisting knn custome methods.

VijayanB committed:
Add new methods for whitelisting

Added new methods which will be whitelisted for users to
call in source field for painless scripting.

VijayanB committed:
Create Whitelist text file

Add Extension class and return list of whitelisted method.
Update gradle to add dependencies to painless.

VijayanB commented:
Following methods are whitelisted:
l2Squared
cosineSimilarity
cosineSimilarityOptimized

@vamshin @jmazanec15  we can discuss here if we want to rename the method name for better reach

VijayanB committed:
Remove Test Annotation

test annotation is forbidden

jmazanec15 commented on src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorFieldMapper.java:
```diff
@@ -207,6 +210,12 @@ public Query termQuery(Object value, QueryShardContext context) {
         }
+
+        @Override
+        public IndexFieldData.Builder fielddataBuilder(String fullyQualifiedIndexName, Supplier<SearchLookup> searchLookup) {
```
fielddataBuilder should be fieldDataBuilder based on IndexFieldData

VijayanB replied:
Did you mean method name? if so, it is from parent class. i am overriding it here

jmazanec15 replied:
Oh haha yes you are correct, my mistake

jmazanec15 commented on src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/script/KNNScoringUtil.java:
```diff
@@ -50,7 +90,7 @@ public static float l2Squared(float[] queryVector, float[] inputVector) {
      */
     public static float cosinesimilOptimized(float[] queryVector, float[] inputVector, float normQueryVector) {
-        float dotProduct = 0.0f;
+            float dotProduct = 0.0f;
```
Why is this indented?

VijayanB replied:
ACk. Will apply reformat code in next commit.

jmazanec15 commented on src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/script/KNNScoringUtil.java:
```diff
@@ -64,6 +104,35 @@ public static float cosinesimilOptimized(float[] queryVector, float[] inputVecto
+     * for query vector for each filtered documents
+     * example
+     *  "script": {
+     *         "source": "cosineSimilarity(params.query_vector, docs[field], 1.0) ",
```
This should be cosineSimilarityOptimized, no?

VijayanB replied:
Ack.

VijayanB replied:
addressed your rename comment, hence keeping cosineSimilarity as overloaded.

jmazanec15 replied:
Makes sense

jmazanec15 commented on src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/script/KNNScoringUtil.java:
```diff
@@ -64,6 +104,35 @@ public static float cosinesimilOptimized(float[] queryVector, float[] inputVecto
+     * @param normQueryVector normalized query vector value.
+     * @return cosine score
+     */
+    public static float cosineSimilarityOptimized(
```
I think this interface is vague: cosineSimilarityOptimized does not give a user any information about how this function differs from cosineSimilarity. If cosineSimilarity can be optimized, why even provide cosineSimilarity?

The Cosine similarity is A . B / (||A|| x ||B||). The difference between the two functions is that cosineSimilarityOptimized has the user pass in both the query vector and the magnitude of the query vector. Assuming the query vector is A, ||A|| does not change throughout the query. So time is saved by computing ||A|| separately and then passing it into the function. Also, I think normQueryVector may not be the appropriate term. Normalization refers to the process of making a vector have a magnitude of one.

My proposal is to switch it to the following interface:
```
cosineSimilarity(List <Number> queryVector, KNNVectorScriptDocValues docValues, Number queryVectorMagnitude)
```

VijayanB replied:
Ack

jmazanec15 commented on src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/script/KNNScoringUtil.java:
```diff
@@ -64,6 +104,35 @@ public static float cosinesimilOptimized(float[] queryVector, float[] inputVecto
     }
 
+    /**
+     * Whitelisted method for users that can be used script to avoid repeated calculation of normalization
```
Nit: "Whitelisted cosineSimilarity method that can be used in a script to avoid repeated calculation of normalization for the query vector.

Example:
...
"

VijayanB replied:
Ack

jmazanec15 commented on src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorScriptDocValues.java:
```diff
@@ -0,0 +1,69 @@
+            throw new IllegalArgumentException("no value found for the corresponding doc ID");
+        }
+        BytesRef value = binaryDocValues.binaryValue();
+        Objects.requireNonNull(value);
```
Is this redundant with line 46?

VijayanB replied:
ACk

jmazanec15 commented on src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorScriptDocValues.java:
```diff
@@ -0,0 +1,69 @@
+        ObjectInputStream objectStream = new ObjectInputStream(byteStream);
+        try {
+            return (float[]) objectStream.readObject();
+        }catch (ClassNotFoundException e) {
```
space before catch

VijayanB replied:
Ack. Applying reformat code.

jmazanec15 commented on src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorScriptDocValues.java:
```diff
@@ -0,0 +1,69 @@
+    @Override
+    public void setNextDocId(int docId) throws IOException {
+        if (binaryDocValues.advanceExact(docId)) {
+            docExists = true;
```
Is this thread safe? I am wondering if this is the correct place to determine if a value exists for that doc.

VijayanB replied:
other scriptdocs like BInary, doesn't make this thread safe. Nevertheless, synchronized the methods to make sure we will not have race conditions.

jmazanec15 replied:
I see. Refer to comment below about count

jmazanec15 commented on src/test/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorDVLeafFieldDataTests.java:
```diff
@@ -0,0 +1,91 @@
+        expectThrows(UnsupportedOperationException.class,
+                () -> leafFieldData.getBytesValues());
+    }
+}
```
nit: add endline

VijayanB replied:
ack

jmazanec15 commented on src/test/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorIndexFieldDataTests.java:
```diff
@@ -0,0 +1,93 @@
+        expectThrows(UnsupportedOperationException.class,
+                () -> indexFieldData.newBucketedSort(null, null, null, null, null, null, 0, null));
+    }
+}
```
nit: add endline

VijayanB replied:
Ack

jmazanec15 commented on src/test/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorScriptDocValuesTests.java:
```diff
@@ -0,0 +1,85 @@
+        expectThrows(UnsupportedOperationException.class,
+                () -> scriptDocValues.get(0));
+    }
+}
```
add endline

VijayanB replied:
Ack

jmazanec15 reviewed (commented):
A few comments. Still need to dig into the test cases more. One note, once #222 is resolved, we will need to add a section on this.

VijayanB committed:
Fixed code review comments
1. Made KNNScriptDocValues thread safe
2. reformat code
3. rename cosinesimiloptimized.

VijayanB committed:
Remove nonnull check

VijayanB committed:
Fixed formatting errors

VijayanB committed:
Add Integration test

Add painless scripting score to verify whether following methods
are available for users.
1. l2Squared
2. cosineSimilarity(queryVector,doc[field])
3. cosineSimilarity(queryVector,doc[field],normalizedVector)

jmazanec15 commented on src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorScriptDocValues.java:
```diff
@@ -0,0 +1,72 @@
+
+    public synchronized float[] getValue() throws IOException {
+        if (!docExists) {
+            throw new IllegalArgumentException("no value found for the corresponding doc ID");
```
Why throw IllegalArgumentException? It seems like it maybe should be an IllegalStateException.

VijayanB replied:
Ack

jmazanec15 commented on src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorScriptDocValues.java:
```diff
@@ -0,0 +1,72 @@
+import java.io.ObjectInputStream;
+
+// This class is thread safe, since docExists is synchronized at an instance level
+public final class KNNVectorScriptDocValues extends ScriptDocValues<float[]> {
```
Should this extend BinaryScriptDocValues<float[]>? Here is the [code](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/index/fielddata/ScriptDocValues.java#L440).

I think if we go this route, we do not have to have docExists boolean.

VijayanB replied:
Thats a good point. That was my first attempt, but BinaryScriptDocValues is not public, hence we can't extend it.

VijayanB replied:
Thats a good point. That was my first attempt, but BinaryScriptDocValues is not public, hence we can't extend it.

jmazanec15 replied:
Ah, I see. That makes sense. In terms of docExists, I think we should follow the convention of using [count](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/index/fielddata/ScriptDocValues.java#L493) instead. What do you think?

VijayanB replied:
We can do that too, but i kept consistent with our other implementation as here
https://github.com/opendistro-for-elasticsearch/k-NN/blob/master/src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/script/KNNScoreScript.java#L138

jmazanec15 replied:
You are right. We probably don't need to make that change now.

jmazanec15 commented on src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorDVLeafFieldData.java:
```diff
@@ -0,0 +1,49 @@
+
+import java.io.IOException;
+
+public class KNNVectorDVLeafFieldData implements LeafFieldData {
```
Should this extend [AbstractBinaryDVLeafFieldData](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/index/fielddata/plain/AbstractBinaryDVLeafFieldData.java)?

VijayanB replied:
same as above. The class is not public.

jmazanec15 commented on src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/script/KNNScoringUtil.java:
```diff
@@ -34,18 +37,55 @@
+        float[] knnDocVector;
+        try {
+            knnDocVector = docValues.getValue();
+        } catch (Exception e) {
```
Should we catch the specific exception here?

VijayanB replied:
getValue can throw RuntimeException & IOException, i see three option here
1) Let docValues throw exception as it is and catch all individual exception here and return Float.Min like now,
2) Update getValue to throw only RuntimeException, and catch only RuntimeException here
3) Just update catch (Exception e) to catch ( RuntimeException | IOException e )

i don't see any difference in all three since at the end outcome is same, but if you have strong opinion on one vs other, i can change it . if you see any other option i can make change it as well.

jmazanec15 replied:
Makes sense, Option 1 us good with me

jmazanec15 commented on src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/script/KNNScoringUtil.java:
```diff
@@ -34,18 +37,55 @@
      *
-     * @param queryVector query vector
-     * @param inputVector input vector
+     * @param queryVector     query vector
```
Out of curiosity, any reason for these tabs here?

VijayanB replied:
![Screen Shot 2020-12-16 at 10 27 38 AM](https://user-images.githubusercontent.com/11067894/102390506-555a9000-3f89-11eb-9605-54773c1ce1b9.png)

jmazanec15 replied:
Oh okay, makes sense. Yes, in the future I think we should use some kind of linter to keep style consistent.

VijayanB replied:
i agree.

jmazanec15 commented on src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/script/KNNScoringUtil.java:
```diff
@@ -64,6 +104,35 @@ public static float cosinesimilOptimized(float[] queryVector, float[] inputVecto
+     *
+     * @param queryVector     query vector
+     * @param docValues       script doc values
+     * @param normQueryVector normalized query vector value.
```
The other part of my comment suggested changing this name to queryVectorMagnitude. normQueryVector makes it seem like the user should pass in the normalized query vector, not the magnitude of the query vector.

VijayanB replied:
I see. I will update it, on side note, i don't think user will see this variable name since they will only see datatype.

jmazanec15 commented on src/test/java/com/amazon/opendistroforelasticsearch/knn/KNNRestTestCase.java:
```diff
@@ -418,15 +419,15 @@ protected void deleteKnnDoc(String index, String docId) throws IOException {
-                       .field(settingKey, value)
-                       .endObject()
-                       .endObject();
+                .startObject()
```
Which style tool did you use for formatting? I think it might make sense to add some kind of styling framework to the code so that we stay consistent (I am not suggesting adding it in this PR, but it might make sense to add in the future).

VijayanB replied:
i am using the one which comes with Intellij

jmazanec15 replied:
Refer to above

jmazanec15 reviewed (commented):
I think test cases look pretty good. I added a few additional comments in implementation.

VijayanB committed:
Fixed review comments

1. updated exception type.
2. Added test

jmazanec15 reviewed (approved):
All looks good to me!

vamshin commented on src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorScriptDocValues.java:
```diff
@@ -0,0 +1,72 @@
+
+    @Override
+    public void setNextDocId(int docId) throws IOException {
+        synchronized (this) {
```
do we need ```synchronized``` here?

VijayanB replied:
I added thread safe to prevent any race condition based on feedback, since docExists setting is happening in one place and reading is happening on another place. If there is not threat with race condition then definitely we can remove this. At the end, it comes to do we need to make this instance thread safe or not? What do you think?

vamshin replied:
Good point. Do you have pointers to any other base classes extending ``` ScriptDocValues<>``` doing synchronization? My understanding is each search request should have its own instance and documents are iterated sequentially so synchronization should not be required. If you see any base class doing this we can definitely consider.

VijayanB replied:
Removed synchronization.

vamshin commented on src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorScriptDocValues.java:
```diff
@@ -0,0 +1,72 @@
+
+    @Override
+    public int size() {
+        synchronized (this) {
```
Do we need ```synchronized``` here?  Avoid synchronization in the places not needed as it could hamper performance.

VijayanB replied:
I added so that docExists is synchronized. Since it is just one instruction, i think i can remove this.

vamshin replied:
Yes please. I have not seen other places doing this. We can remove this.

VijayanB replied:
Removed

vamshin commented on src/test/java/com/amazon/opendistroforelasticsearch/knn/plugin/script/PainlessScriptScoringIT.java:
```diff
@@ -0,0 +1,124 @@
+
+    private Map<String, Float[]> getL2TestData() {
+        Map<String, Float[]> data = new HashMap<>();
+        data.put("1", new Float[]{6.0f, 6.0f});
```
How about we have a document without the vector and then assert that particular doc comes at the end of the result? Same for both ```CosineTestData``` as well.

VijayanB replied:
Added test data.

VijayanB committed:
Fix review comments

1. Requires dimension must be consistent between query vector and input vector
2. Remove thread safe for KNNScriptDocValues
3. Throw exception if document doesn't contains vector field.

VijayanB committed:
Fixed lint errors

VijayanB committed:
Add test case to check size

To avoid exception, user's can use size method to validate.

VijayanB committed:
Extract constant

VijayanB committed:
Update error message

Include error message with additional details on how to fix the
problem.

VijayanB committed:
Update error message

Add field name as part of error message

jmazanec15 reviewed (approved):
LGTM!

VijayanB committed:
Refactor ScoreScirpt

Use scriptdoc to extract values from segment.

VijayanB committed:
Refactor ScoreScript

Use KNNScriptDocValues to retrieve float[] from the
document instead of explicity retrieval.

vamshin reviewed (approved):
LGTM! Thanks for adding painless support. This will help users to have more customization on the scoring.
